### ISSUE TITLE
iOS Annotation view sticked at the left of the screen when dragging map

### ISSUE THREAD

**Issue by evgzor** [mapbox/mapbox-gl-native#8603]
Hello,

I have a strange behavior on 3.5.0 release for the map with  annotations like: 
![screen shot 2017-03-31 at 18 40 21](https://cloud.githubusercontent.com/assets/5227456/24557756/86d0bf12-1641-11e7-8e71-f72cf846d347.png)
Please also look attached video.
I just drag the map to the left and it appears new annotation from the left side. Before 3.3.4 I did not have such issues. Probably u had similar problem before?

**Platform:**
iOS 10.2
**Mapbox SDK version:**
3.5.0
### Steps to trigger behavior

 1. Load map with annotations with custom pictures
 2. Drug map to left corner so like annotation will be out of map edge.
 3. Observe the result

### Expected behavior
No side effect from dragging and all annotations on the same place.
### Actual behavior
Appears annotation which was out of map edge.
[MapBox issue.zip](https://github.com/mapbox/mapbox-gl-native/files/886187/MapBox.issue.zip)

**Comment by fabian-guerra** [comment]
Hi @evgzor sorry to hear you are having issues with the latest iOS SDK release, if you can provide a code example that help us reproduce the problem, would likely help us solve this issue faster.

Thank you.

**Comment by evgzor** [comment]
Hi @fabian-guerra I attached code source
[AnnotationViewMultipleExample.m.zip](https://github.com/mapbox/mapbox-gl-native/files/887602/AnnotationViewMultipleExample.m.zip) from mapbox example project with simulated problem. Please have a look. If u need full project u may checkout from https://github.com/rmnblm/Mapbox-iOS-Examples.git and replace modified file.

**Comment by fabian-guerra** [comment]
Hi @evgzor thank you for sending an example, I'm able to see the problem. One more question did you updated the SDK from 3.3.4 to 3.5.0?

Thanks again!

**Comment by evgzor** [comment]
Hi @fabian-guerra right, I updated from 3.3.4 to 3.5.0 for getting MGLStyle API. However, I observed the same problem on 3.4.2 SDK. Do u still need more info from my side?

**Comment by fabian-guerra** [comment]
Hi @evgzor we have found the issue. When you create a custom `MGLAnnotationView` you need to set the `frame`, you can see _something_ because the `UIImageView` automatically sizes itself to fit the `UIImage`.

You can fix this by setting the frame e.g `annotationView.frame = imageView.frame;`

/cc @boundsj

**Comment by boundsj** [comment]
https://github.com/mapbox/mapbox-gl-native/issues/5469 would simplify this use case by providing an `image` property that could be set to a UIImage. `MGLAnnotationView` would add the image and size itself appropriately (the work that is currently being _almost_ done in the example noted here)

**Comment by evgzor** [comment]
@fabian-guerra if I set frame like (0,0,1,1) the issue does not fix and if i set annotationView.frame = imageView.frame; -- I get 
<img width="376" alt="screenshot 2017-04-03 22 04 37" src="https://cloud.githubusercontent.com/assets/5227456/24626326/9a77830a-18b9-11e7-9f22-344d6557914f.png"> and it make problem bigger.
@boundsj I have implemented own image property for 3.3.4 - and it worked ok till but it works in a little bit different way. I cannot easily use to new approach. It very difficult adopt application when u start to change behavior. From my point of view it's a bug if I cannot use annotation with zero frames  (as it was before). You closed it without getting any feedback. 
If it's easier for you, ok ...

**Comment by boundsj** [comment]
I'm sorry the workaround in https://github.com/mapbox/mapbox-gl-native/issues/8603#issuecomment-291231223 did not work for you @evgzor. 

In 3.4.0, we made some [performance improvements](https://github.com/mapbox/mapbox-gl-native/pull/5987) to the way annotation views get updated when the map moves. A side effect of the improvements is that we no longer handle this case where the `MGLAnnotationView` is not large enough to contain its subviews.

We could add logic to the SDK to re-enable support for this by special casing invisible `MGLAnnotationViews` and always moving them but that would not solve for `MGLAnnotationViews` that are still visible but smaller than their content. Another option might be to add logic to dig through all of the subviews that an `MGLAnnotationView` has and determine a sort of `sizeThatFits` and use that for the view update calculation. However, I'm concerned about introducing another performance degradation by doing that.

I can say that simply using `(0,0,1,1)` as the frame won't work (although it might if your `UIImageView` were much smaller).

I'm not sure why setting `MGLAnnotationView`'s width and height to be big enough to contain the `UIImageView` does not work for you. When I try that with your example it appears to work fine.

That is:

```Objective-C
UIImage *image = [UIImage imageNamed:@"lighthouse"];
UIImageView* imageView = [[UIImageView alloc] initWithImage:image];
CGRect annotationViewFrame = CGRectMake(0, 0, CGRectGetWidth(imageView.frame), CGRectGetHeight(imageView.frame));
annotationView.frame = annotationViewFrame;
[annotationView addSubview:imageView];
```

Looks like:

![annotation](https://cloud.githubusercontent.com/assets/63186/24631709/2ded6e90-1876-11e7-97db-13e6014b8308.gif)

Can you please show a code sample where you set the annotation view's width and height to be the same as the image view and say more about why this make the problem worse for you?

**Comment by fabian-guerra** [comment]
Hi @evgzor sorry that https://github.com/mapbox/mapbox-gl-native/issues/8603#issuecomment-291231223 didn't work for you. Complementing what I said you can use a parent view that fits the `UIView` subview combined with`MGLAnnotationView.centerOffset` e.g. 

`annotationView.centerOffset = CGVectorMake(0, -CGRectGetHeight(imageView.frame)/2.0);`

This will help you fix the offset you see in https://github.com/mapbox/mapbox-gl-native/issues/8603#issuecomment-291244211.

**Comment by evgzor** [comment]
@boundsj and @fabian-guerra for the moment I fixed like that: frame = CGRect(x: 0, y: 0, width: 4, height: 0). and overide centerOffset  self.centerOffset = CGVector(dx: self.centerOffset.dx + self.bounds.width / 2 - .... Its workaround, but for complex annotation view which contains several views (not only image view) I have to use that approach. In general it implemented for SDK which can swith between MKMapView and MapBox map, by toggle  and I have to always alighn high level api.

**Comment by cansurmeli** [comment]
I'm trying to implement a custom MGLAnnotationView on iOS with Swift as well. And this issue seems to be continuing even on SDK 3.5.2.

Besides the annotation locations not updating continuously along with the map view, when the map view is initially launched, the annotations appear after a delay of couple of seconds but not immediately.

**Comment by boundsj** [comment]
This should be fixed in the [Mapbox iOS v3.5.4 release](https://github.com/mapbox/mapbox-gl-native/releases/tag/ios-v3.5.4)

**Comment by kiklenain** [comment]
Hello, I get the same behavior on the v3.7.1 and v3.7.6 when the map is pitched
